components/data-feed-service/service/data-feed-workflow.go

```diff
@@ -0,0 +1,157 @@
+	params := DataFeedWorkflowParams{}
+	log.Debugf("OnStart params %v", params)
+	err := w.GetParameters(&params)
+	if err != nil {
+		return w.Fail(err)
+	}
+
+	/*
+	 * Start the workflow with the poll task to get any nodes that have had client runs
+	 */
+	err = w.EnqueueTask(dataFeedPollTaskName, params)
+	if err != nil {
+		return w.Fail(err)
+	}
+
+	initialPayload := DataFeedWorkflowPayload{}
+	return w.Continue(&initialPayload)
+}
+
+func (e *DataFeedWorkflowExecutor) OnTaskComplete(w cereal.WorkflowInstance, ev cereal.TaskCompleteEvent) cereal.Decision {
+	params := DataFeedWorkflowParams{}
+	err := w.GetParameters(&params)
+	if err != nil {
+		return w.Fail(err)
+	}
+
+	log.Debugf("OnTaskComplete params: %v", params)
+	payload := DataFeedWorkflowPayload{}
+	err = w.GetPayload(&payload)
+	if err != nil {
+		return w.Fail(err)
+	}
+
+	switch ev.TaskName {
+	case dataFeedPollTaskName:
+		taskResults, err := getPollTaskResults(ev)
+		if err != nil {
+			return w.Fail(err)
+		}
+		/*
+		 * Update the workflow params
+		 */
+		params.NodeIDs = taskResults.NodeIDs
+		if len(taskResults.NodeIDs) > 0 {
+			payload.NodeIDs = batchNodeIDs(params.NodeBatchSize, taskResults.NodeIDs)
+			log.Debugf("payload.NodeIDs %v payload.NodeIDsIndex %v", payload.NodeIDs, payload.NodeIDsIndex)
+			params.NodeIDs = payload.NodeIDs[payload.NodeIDsIndex]
+			err = w.EnqueueTask(dataFeedAggregateTaskName, params)
+			if err != nil {
+				return w.Fail(err)
+			}
+		} else {
+			// no nodes have been updated by client run and
+			// no compliance reports have been run in the interval
+			// the workflow is therefore complete
+			payload.TasksComplete = true
+			log.Debugf("No new client data or reports found")
+		}
+		log.Debugf("PollTask %v, tasks complete: %v, time: %v", dataFeedPollTaskName, payload.TasksComplete, time.Now())
+	case dataFeedAggregateTaskName:
+		payload.NodeIDsIndex++
+		if payload.NodeIDsIndex != len(payload.NodeIDs) {
+			params.NodeIDs = payload.NodeIDs[payload.NodeIDsIndex]
+			err = w.EnqueueTask(dataFeedAggregateTaskName, params)
+			if err != nil {
+				return w.Fail(err)
+			}
+		} else {
+			payload.TasksComplete = true
+		}
+		log.Debugf("AggregateTask %v, tasks complete: %v, time: %v", dataFeedAggregateTaskName, payload.TasksComplete, time.Now())
+	}
+
+	log.Debugf("TasksComplete %v, complete: %v", ev.TaskName, payload.TasksComplete)
+	if payload.TasksComplete {
+		return w.Complete()
+	}
+	return w.Continue(&payload)
+
+}
+
+func (e *DataFeedWorkflowExecutor) OnCancel(w cereal.WorkflowInstance, ev cereal.CancelEvent) cereal.Decision {
+	return w.Complete()
+}
+
+func batchNodeIDs(batchSize int, nodeIDs map[string]NodeIDs) []map[string]NodeIDs {
+	numNodes := len(nodeIDs)
+	log.Debugf("number of nodes %v", numNodes)
+	var numBatches int
+	if numNodes < batchSize {
+		numBatches = 1
+	} else {
+		numBatches = numNodes / batchSize
+	}
+	if numNodes%1 > 0 {
```

Comment:
I may be missing something, but isn't this always 0? I've never modded by 1 before, but 
`0%2=0` `1%2=1` `2%2=0`?

Reply:
not sure how that 1 got in there, it's fixed now. Simplified a little and it is numNodes%batchSize.

Example 25 nodes, batch size 10. `numBatches = 2` but 25 nodes requires 3 batches thus if `numNodes%batchSize > 0`, `25%10 = 5` we need another batch